### ISSUE TITLE
Buffer overflow when compiling dictionaries

### ISSUE THREAD

**Issue by feerrenrut** [espeak-ng/espeak-ng#287]
### Background:
While updating the NVDA espeak-ng submodule to commit fb97d1bd7564c2ff7c305cf7fcbdd29132234846 we have run into some problems compiling the dictionaries. The build system was occasionally halting with a python crash. After some investigation I found that a missing new line at the end of 'ar_listx' was causing a buffer overrun, which I have worked around with https://github.com/nvaccess/espeak-ng/commit/0994206f710a4defc1eecfb78ab70ff57c58fcda.

For this problem (missing new line character); assuming there is no technical reason that a new line character must be present, I suggest that the dictionary compilation is modified to accept files with missing new lines. Otherwise, to save time in debugging and accidental newlines I suggest that a missing newline is detected and reported during dictionary compilation.

There still seems to be a crash or sometimes the process runs indefinitely. Further investigation has led me to find that in `LoadDictionary` length is sometimes negative. Trying to match up the signed / unsignedness of 'length' (particularly in `compile_dictlist_end`) I found in some cases it is actually negative when written. On my system I can reliably reproduce this when running through 'compile_dictlist_end()' for the 'bn' dictionary, at hash number 497 we eventually get a length value of -117 (signed) / 4294967179 (unsigned). 

I don't really understand how length comes to be -117. I added some asserts (and then built espeak with `/SDL /RTC1 /MDd /Od /D_DEBUG ` MSVC flags) to make debugging this a little easier, see the full diff of espeak: https://github.com/espeak-ng/espeak-ng/compare/master...nvaccess:fixEspeakCrashDuringDictCompilation

Also the branch of NVDA that is used to build the `fixEspeakCrashDuringDictCompilation` is [updateEspeak-debugBranch](https://github.com/nvaccess/nvda/tree/updateEspeak-debugBranch) specifically commit: https://github.com/nvaccess/nvda/commit/6bf896248798ee041c034372cc0b18efa5c611dc 

### Key points:
- Missing new line at the end of 'ar_listx' causes buffer overrun
- Signed / unsigned mismatch of `length` in compile_dictlist_end / `LoadDictionary` 
- This value appears to be either too large or negative, indicating a problem.

### To reproduce locally:
- On a windows machine (with the required NVDA dependencies)
- Clone https://github.com/nvaccess/nvda repository
- Checkout 6bf896248798ee041c034372cc0b18efa5c611dc
- Build nvda with `python scons.py source`

**Comment by rhdunn** [comment]
Thanks for the investigation you have done here, I have used that to identify and hopefully fix the problem. I have merged your fix for dictionary files with no trailing newlines, but have not merged the rest of that branch.

In the emoji for `bn`, some of the entries were longer than 128 bytes, so were overflowing `dict_line`. I added a check for that overflow in d5d980862e3a487db22f8ef7e761f0890f1b6bed and fixed that overflow in 2a00ca79f635d8e7979ad09846901d59a4ea4ee0, resulting in the issue you are seeing.

The espeak code stores the length of the line in the first byte, and is using `char` to point to that line buffer. If `char` is signed, this will result in negative values like you are seeing. I have addressed that in ebfa320956169e3419234b72fee51bd596867661 by casting the buffer to `uint8_t` before reading the value.
